src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py

```diff
@@ -1670,7 +1673,11 @@ def with_query_block(
         # query generation stage.
         queries = child.queries[:-1] + [Query(sql=new_query)]
         # propagate the cte table
-        referenced_ctes = {with_query_block}.union(child.referenced_ctes)
+        referenced_ctes = child.referenced_ctes.copy()
+        if with_query_block in referenced_ctes:
+            referenced_ctes[with_query_block] += 1
```

Comment:
once you have the util funciton, you can just call it with {with_query_block: 1}